cpu/gd32v/periph/i2c.c

```diff
@@ -189,227 +208,522 @@ void i2c_release(i2c_t dev)
+    assert(dev < I2C_NUMOF);
+    assert(data);
+    assert(length);
+
+    DEBUG("i2c_write_bytes, dev=%d len=%d flags=%02x\n", dev, length, flags);
+
+    int res;
+
+    _i2c_dev[dev].state = I2C_OK;
+
+    /*  if I2C_NOSTART is not set, send START condition and ADDR */
+    if (!(flags & I2C_NOSTART)) {
+        res = _i2c_start_cmd(dev);
+        if (res != 0) {
+            return res;
+        }
+
+        /* address handling */
+        if (flags & I2C_ADDR10) {
+            /* prepare 10 bit address bytes */
+            uint8_t addr10[2];
+            addr10[0] = 0xf0 | (addr & 0x0300) >> 7;
+            addr10[1] = addr & 0xff;
+            /* send ADDR without read flag */
+            res = _i2c_addr_cmd(dev, addr10, 2);
         }
-        if ((sr1 & I2C_STAT0_LOSTARB_Msk) || (sr1 & I2C_STAT0_BERR_Msk)) {
-            DEBUG("[i2c] is_sr1_mask_set: arb lost or bus ERROR_FLAGS\n");
-            i2c->STAT0 &= ~ERROR_FLAGS;
-            _stop(i2c);
-            return -EAGAIN;
+        else {
+            /* send ADDR without read flag */
+            uint8_t addr7 = addr << 1;
+            res = _i2c_addr_cmd(dev, &addr7, 1);
         }
-        if (sr1 & mask) {
-            i2c->STAT0 &= ~ERROR_FLAGS;
-            return 0;
+        if (res != 0) {
+            return res;
+        }
+    }
+
+    /* read STAT0 followed by reading STAT1 to clear the ADDSEND flag */
+    i2c_config[dev].dev->STAT0;
+    i2c_config[dev].dev->STAT1;
+
+    /* send data */
+    uint8_t *buffer = (uint8_t*)data;
+
+    for (size_t i = 0; i < length; i++) {
+        res = _i2c_write_cmd(dev, buffer[i]);
+        if (res != 0) {
+            return res;
         }
     }
-    /*
-    * If timeout occurs this means a problem that must be handled on a higher
-    * level.  A SWRST is recommended by the datasheet.
-    */
-    i2c->STAT0 &= ~ERROR_FLAGS;
-    _stop(i2c);
-    return -ETIMEDOUT;
+
+    /* if I2C_NOSTOP flag is not set, send STOP condition */
+    if (!(flags & I2C_NOSTOP)) {
+        /* send STOP condition */
+        res = _i2c_stop_cmd(dev);
+        if (res != 0) {
+            return res;
+        }
+    }
+
+    return _wait_for_bus(dev);
 }
 
-static int _stop(I2C_Type *i2c)
+void _i2c_transfer_timeout(void *arg)
 {
-    /* send STOP condition */
-    DEBUG("[i2c] stop: Generate stop condition\n");
-    i2c->CTL0 &= ~(I2C_CTL0_ACKEN_Msk);
-    i2c->CTL0 |= I2C_CTL0_STOP_Msk;
-    uint16_t tick = TICK_TIMEOUT;
-    while ((i2c->CTL0 & I2C_CTL0_STOP_Msk) && tick--) {}
-    if (!tick) {
-        return -ETIMEDOUT;
-    }
-    DEBUG("[i2c] stop: Stop condition succeeded\n");
-    if (_wait_for_bus(i2c) < 0) {
-        return -ETIMEDOUT;
+    i2c_t dev = (i2c_t)(uintptr_t)arg;
+
+    /* set result to timeout */
+    _i2c_dev[dev].state = I2C_TIMEOUT;
+
+    /* wake up the thread that is waiting for the results */
+    mutex_unlock(&_i2c_dev[dev].irq_lock);
+}
+
+static void _wait_for_irq(i2c_t dev)
+{
+#if defined(MODULE_ZTIMER_USEC)
```

Comment:
Timeouts are realized either
- by means of `ztimer_usec` timers, if the `ztimer_usec` module is already activated,
- by `ztimer_msec` timer if the `ztimer_msec` module is already activated, or
- not at all.
The driver itself does not pull a timer module by itself.

However, the use of `ztimer_msec` seems to make no sense at all, since it uses the RTC, for which a write operation to an RTC register requires at least 3 RTC clock cycles (i.e. about 91 us). This means that setting and removing the timer again takes about 180 us in total, which is about twice the duration for sending/receiving a byte at 100 kHz clock. As a result, there seems to be an idle phase between the sent and received bytes, which is about the duration of a sent or received byte and thus reduces the speed of the I2C transmission to about half, see the image below.

Even worse, during this time (the entire 180 us) the MCU is constantly polling the RTC's status register to determine when the write access for the timer set/unset is complete. So instead of letting the MCU sleep while waiting for an interrupt, the MCU is in a waiting state the whole time, which makes the interrupt-driven I2C driver completely absurd

![gd32v_i2c_irq](https://user-images.githubusercontent.com/31932013/218555348-3d77f52b-ff3e-432a-adca-de15d9790017.png)

With timeout handling using `ztimer_usec` it looks as it should:

![gd32v_i2c_polling](https://user-images.githubusercontent.com/31932013/218555400-6234357e-03dc-419d-b7d7-3cc34ffb44f6.png)

So the question is whether we remove the timeout handling to avoid using `ztimer_usec`, or whether we enable `ztimer_usec` for the I2C driver. If we use timeout handling, it makes only sense with `ztimer_usec`.

Reply:
> I think `ztimer_usec` is fine.
> There is no requirement to avoid it except for long-running timers. If you only need a short delay there is nothing wrong with `ztimer_usec`.

But doesn't this mean that the timer peripheral will run all the time even if I2C is not used and no timer is set, at least if the `ztimer_timer_ondemand` module is not used?

Reply:
I didn't follow the discussion there tbh, but afaik the (long term) goal is that the timer should be only running if it has alarms scheduled or if it was acquired somehow to measure time.

Since we are not there yet (there is plenty of code that relies on the timer to be always running to measure time intervals) this is indeed the case, but your code should not bend over backwards to accommodate a temporary (let's hope so!) shortcoming of the ztimer implementation.

It already does the right thing by scheduling an alarm, so once we flip the switch (Is `ztimer_timer_ondemand` that switch?) it will just behave as it should and only enable the clock when needed.